include/sparrow/arrow_interface/arrow_array_stream/private_data.hpp

```diff
@@ -0,0 +1,126 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <queue>
+#include <ranges>
+#include <string>
+
+#include <sparrow/c_interface.hpp>
+
+namespace sparrow
+{
+    class arrow_array_stream_private_data
+    {
+    public:
+
+        arrow_array_stream_private_data() = default;
+
+        ~arrow_array_stream_private_data()
+        {
+            if (m_schema != nullptr)
+            {
+                if (m_schema->release != nullptr)
+                {
+                    m_schema->release(m_schema);
+                }
+                delete m_schema;
+                m_schema = nullptr;
+            }
+
+            while (!m_arrays.empty())
+            {
+                ArrowArray* array = m_arrays.front();
+                if (array != nullptr)
+                {
+                    if (array->release != nullptr)
+                    {
+                        array->release(array);
+                    }
+                    delete array;
+                }
+                m_arrays.pop();
+            }
+        }
+
+        void import_schema(ArrowSchema* out_schema)
+        {
+            if (m_schema != nullptr)
+            {
+                if (m_schema->release != nullptr)
+                {
+                    m_schema->release(m_schema);
+                }
+                delete m_schema;
+                m_schema = nullptr;
+            }
+
+            m_schema = out_schema;
+        }
+
+        [[nodiscard]] ArrowSchema* schema()
+        {
+            return m_schema;
+        }
+
+        [[nodiscard]] const ArrowSchema* schema() const
+        {
+            return m_schema;
+        }
+
+        template <std::ranges::input_range R>
+            requires std::same_as<std::ranges::range_value_t<R>, ArrowArray*>
+        void import_arrays(R&& arrays)
+        {
+            for (auto&& array : arrays)
+            {
+                m_arrays.push(array);
+            }
+        }
+
+        void import_array(ArrowArray* array)
+        {
+            m_arrays.push(array);
+        }
+
+        ArrowArray* export_next_array()
+        {
+            if (m_arrays.empty())
+            {
+                return new ArrowArray{};
+            }
+
+            ArrowArray* array = m_arrays.front();
+            m_arrays.pop();
+            return array;
+        }
+
+        [[nodiscard]] const std::string& get_last_error_message() const
+        {
+            return m_last_error_message;
+        }
+
+        void set_last_error_message(std::string_view message)
+        {
+            m_last_error_message = message;
+        }
+
+    private:
+
+        ArrowSchema* m_schema = nullptr;
+        std::queue<ArrowArray*> m_arrays{};
```

Comment:
I'm curious, why don't we use "managed" wrappers for the `ArrowArray*` elements? This would avoid hand-writing a bunch of `release` calls and `delete`s.

Perhaps just `std::unique_ptr` with a custom deleter...  And when you want to transfer ownership to a third-party library, you just call `std::unique_ptr::release`.

Same question about `ArrowSchema*` of course.